### PR TITLE
feat(trace): experimental trace generation

### PR DESCRIPTION
**Pull request by dgozman**
This introduces basic tracing behind PWTRACE environment variable:
- capturing network resources;
- capturing snapshots at the start of most actions;
- manually capturing snapshots by test runner when the test fails.

### CODE REVIEW

File: src/trace/snapshotter.ts [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+    }).join(',');
+  };
+
+  const processStyle = (base: string, style: string) => {
+    // TODO: this is wrong.
+    for (const match of style.matchAll(/(url\s*\(['"]?)([^)'"]*)(['"]?\))/g))
+      addResource(base, match[2]);
+  };
+
+  // TODO: handle CSSOM.
```
Comment (pavelfeldman) [inline src/trace/snapshotter.ts:326]: What is the point in capturing style resources if we don't capture CSSOM?

File: src/trace/snapshotter.ts [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+      processStyle(node.baseURI, node.textContent || '');
+
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const element = node as Element;
+      builder.push('<');
+      builder.push(element.nodeName);
+      for (let i = 0; i < element.attributes.length; i++) {
+        const name = element.attributes[i].name;
+        let value = element.attributes[i].value;
+        if (name.toLowerCase() === 'value' && (element.nodeName === 'INPUT' || element.nodeName === 'TEXTAREA'))
```
Comment (pavelfeldman) [inline src/trace/snapshotter.ts:362]: Extract it? Looks expensive to call toLowerCase on every attribute of every node 7 times?

**Review by pavelfeldman — APPROVED** [review]

File: src/browser.ts [hunk 3]
```diff
@@ -49,10 +51,12 @@ export abstract class BrowserBase extends EventEmitter implements Browser {
   readonly _options: BrowserOptions;
   private _downloads = new Map<string, Download>();
   _defaultContext: BrowserContextBase | null = null;
+  _traceController?: TraceController;
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/browser.ts:54]: Yep, in a follow up.

File: src/trace/browserContextTracer.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  viewportSize?: types.Size,
+};
+
+export type ContextDestroyedTraceEvent = {
+  type: 'context-destroyed',
+  timestamp: number,
+  contextId: string,
+};
+
+export type ResourceTraceEvent = {
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/trace/browserContextTracer.ts:42]: Done.

File: src/trace/browserContextTracer.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  type: 'resource',
+  timestamp: number,
+  contextId: string,
+  resourceId: string,
+  url: string,
+  contentType: string,
+  headers: types.Headers,
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/trace/browserContextTracer.ts:49]: responseHeaders

File: src/trace/browserContextTracer.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+    const event: ContextCreatedTraceEvent = {
+      type: 'context-created',
+      timestamp: helper.monotonicTime(),
+      contextId: this._contextId,
+      isMobile: !!this._context._options.isMobile,
+      deviceScaleFactor: this._context._options.deviceScaleFactor || 1,
+      viewportSize: this._context._options.viewport || undefined,
+    };
+    this._contextEventPromise = controller.appendTraceEvent(event);
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/trace/browserContextTracer.ts:87]: Done.

File: src/trace/snapshotter.ts [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+  frameSnapshot.resources.push({ resourceId: resource.resourceId, url });
+
+  const references = await resource.references;
+  await Promise.all(references.map(reference => addResource(savedResourceSet, frameSnapshot, frame, reference)));
+}
+
+export async function captureSnapshot(progress: Progress, savedResourceSet: SavedResourceSet, page: Page): Promise<PageSnapshot | null> {
+  const removeNoScript = !(page.context() as BrowserContextBase)._options.javaScriptEnabled;
+
+  const visit = async (frame: frames.Frame): Promise<FrameSnapshot | null> => {
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/trace/snapshotter.ts:163]: Done.

File: src/trace/browserContextTracer.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    this._context = context;
+    this._contextId = 'context' + (++contextCounter);
+    this._eventListeners = [
+      helper.addEventListener(this._context, Events.BrowserContext.Page, this._onPage.bind(this)),
+    ];
+
+    const event: ContextCreatedTraceEvent = {
+      type: 'context-created',
+      timestamp: helper.monotonicTime(),
+      contextId: this._contextId,
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/trace/browserContextTracer.ts:82]: We'll add them when we need them.

File: src/trace/snapshotter.ts (hunk 1 above)
Comment (dgozman) [reply src/trace/snapshotter.ts:326]: Yeah, I'll look into it.

**Pull request closed without merging**